Fix memory leak on js! block

yjh0502 opened the pull request:
Fix https://github.com/koute/stdweb/issues/148

It seems that refcount for JS reference is incremented before serialized to wasm buffer[1], so refcount will be incresed twice if `Reference::from_raw_unchecked` is used on `deserialize`.

[1] https://github.com/koute/stdweb/blob/85f7aa484f3be9808b992b9258bd85fa588e05b1/src/webcore/runtime.js#L303-L311

koute commented:
Thanks for the PR!

Hmm... it'd be great to have a unit test for this so that we can avoid reintroducing something like this by mistake in the future. Would you be up to writing one?

kuviman commented:
As mentioned in the issue, this breaks stable ids, here's code that breaks after this PR.

```rust
#[macro_use]
extern crate stdweb;

#[test]
fn test() {
    js! { Module.test = {}; }
    let a = js! { return Module.test; }.as_reference().unwrap().as_raw();
    let b = js! { return Module.test; }.as_reference().unwrap().as_raw();
    assert_eq!(a, b);
}
```

This can be fixed though by not deleting from `ref_to_id_map`.

yjh0502 commented:
@koute I added testcases. A testcases relies on runtime details, like `Module.STDWEB_PRIVATE.id_to_refcount_map`. Is it okay?

@kuviman I changed the patch to retain reference ids.

koute commented on src/webcore/value.rs:
```diff
@@ -1291,4 +1297,50 @@ mod tests {
+            let obj = js! { return new Object(); };
+            obj.as_reference().unwrap().as_raw()
+        };
+        assert_noref(refid);
```
Could you make this more explicit? e.g.:

```
let value = js! { return new Object(); }.into_reference().unwrap();
let refid = value.as_raw();
assert_eq!( is_known_reference(refid), true );
mem::drop( value );
assert_eq!( is_known_reference(refid), false );
```

Basically, to explicitly show that dropping the `Reference` will also remove it from the map.

yjh0502 replied:
I changed the patch to use `drop` explicitly.

koute commented on src/webcore/value.rs:
```diff
@@ -1291,4 +1297,50 @@ mod tests {
+        js! { Module.__test = {}; }
+    }
+
+    fn assert_noref(refid: i32) {
```
I think this would be better as a function which returns `bool` - then you could test the inverse. (Also, perhaps a more descriptive name? `is_known_reference`? These are the tests, so we don't care if it's long.)

yjh0502 replied:
Fixed.

koute commented on src/webcore/value.rs:
```diff
@@ -1291,4 +1297,50 @@ mod tests {
+        let c = js! { return Module.__test; }.as_reference().unwrap().as_raw();
+        assert_ne!(a, c);
+
+        js! { Module.__test = {}; }
```
Shouldn't this be `delete`d at the end?

yjh0502 replied:
Fixed. Thanks.

koute commented on src/webcore/value.rs:
```diff
@@ -1291,4 +1297,50 @@ mod tests {
+        js! { delete Module.__test; }
+        js! { Module.__test = {}; }
+        let c = js! { return Module.__test; }.as_reference().unwrap().as_raw();
+        assert_ne!(a, c);
```
Is there any point in reassigning the `Module.__test` now? This might just as well be replaced with `js! { return {}; }`. (Since the fact that we're returning `Module.__test` doesn't matter; what matters here is that we `deleted`d the original `{}` and we returned a new `{}`.)

yjh0502 replied:
I agree.

koute commented:
Yeah, depending on the internals in tests is fine.

Also, you might want to get rid of those `.swp` files you've accidentially `git add`ed. (: (You can add `*.swp` to `.gitignore` if you want.)

yjh0502 commented:
@koute Thanks for the review. I updated the patch, and also rebased the PR to `master` branch.

koute commented:
LGTM! Although I can't merge this as there is still `.runtime.js.swp` file added. (:

So please remove that and it'll be good to go!

yjh0502 commented:
@koute I missed that one, sorry :(.

koute commented:
No problem!

Thanks! (:
